### PR TITLE
Add new template to disallow using the latest tag in a container image

### PR DESCRIPTION
**Pull request by arapulido**
It is fairly common policy in Kubernetes to disallow using the `latest` tag in some clusters (especially staging and production).

This PR adds a new template for a policy that will check that tags for images are included, and that the tag is not `latest`

### CODE REVIEW

File: library/general/disallowlatest/template.yaml
Comment (maxsmythe) [inline library/general/disallowlatest/template.yaml:19]: Performing a suffix match might be safer here. Either that or we should anchor the regex to the end of the string using "$", though we should strip whitespace if we do that.

File: library/general/disallowlatest/template.yaml
Comment (maxsmythe) [inline library/general/disallowlatest/template.yaml:26]: The `contains()` builtin is probably best to use here:

https://www.openpolicyagent.org/docs/latest/policy-reference/#strings

File: library/general/disallowlatest/template.yaml
Comment (maxsmythe) [inline library/general/disallowlatest/template.yaml:32]: We can avoid duplicating logic by creating a list of all containers before iterating over them.

https://github.com/open-policy-agent/gatekeeper-library/blob/ce24dd6802b8c845f80a27731b9095cc0864726f/src/pod-security-policy/read-only-root-filesystem/src.rego

is an example of how this has been done elsewhere.

**Review by maxsmythe — CHANGES_REQUESTED** [review]
Thanks for the PR!

This fills a significant gap.

I'm wondering if we could adjust the constraint template to take a list of disallowed tags as a parameter. That would allow this template to cover both the "latest" use case and any other tags users may want to block.

**Comment by arapulido** [comment]
Thanks a lot for the quick review! I will work on those changes asap

File: library/general/disallowedtags/template.yaml
Comment (maxsmythe) [inline library/general/disallowedtags/template.yaml:28]: Should we create an object that maps disallowed tags so that we don't need to scroll through the list every time when there are multiple containers?

@tsandall is there some way to make sure that a map of tags would be created at the entry point of evaluating the rule so we guarantee we aren't paying the cost of construction * the # of images?

In practice, this may be fine as the # of images and the # of disallowed tags may not be very large.

Comment (maxsmythe) [inline library/general/disallowedtags/template.yaml:28]: We should add a `:` prefix to the tag to make sure we are denying the full tag and not just a tag suffix.

For example: `myimage:shangrilatest` would be blocked under the current constraint as-written.

**Review by maxsmythe — CHANGES_REQUESTED** [review]

File: library/general/disallowedtags/template.yaml
Comment (arapulido) [reply library/general/disallowedtags/template.yaml:28]: Very true! I will fix this.

Comment (arapulido) [reply library/general/disallowedtags/template.yaml]: Added a commit to address this. I will wait on the response on the object creation to address the other one.

**Review by maxsmythe — APPROVED** [review]
LGTM

Thanks for this! Waiting on second reviewer now.

**Comment by sozercan** [comment]
Hi @arapulido 👋 thanks for the PR! Can you copy rego and add tests in `src`? example: https://github.com/open-policy-agent/gatekeeper-library/tree/master/src/general/allowedrepos

File: library/general/disallowedtags/samples/container-image-must-not-have-latest-tag/example_disallowed2.yaml
Comment (sozercan) [inline library/general/disallowedtags/samples/container-image-must-not-have-latest-tag/example_disallowed2.yaml:12]: nit: newline

**Comment by arapulido** [comment]
> Hi @arapulido 👋 thanks for the PR! Can you copy rego and add tests in `src`? example: https://github.com/open-policy-agent/gatekeeper-library/tree/master/src/general/allowedrepos

Hello @sozercan 👋  Nice to see you around :) Sure, I will work on those

**Review by sozercan — APPROVED** [review]
LGTM

**Pull request merged**
